### ISSUE TITLE
Texture coordinates incorrect

### ISSUE THREAD

**Issue by blair-ahlquist** [Whinarn/MeshDecimator#13]
In `RemoveVertexPass`, `CalculateError(ref vertices[i0], ref vertices[i1], out p, out pIndex)` returns a `p` that comes from neither of vertices corresponding to `i0` or `i1`, so `MergeVertexAttributes` does not interpolate attributes properly.

**Comment by Whinarn** [comment]
The fact that the resulting _p_ is different from both vertices _i0_ and _i1_ is entirely expected, since it could be an adjustment for preventing the mesh to loose most of its shape during mesh decimation.

The resulting _p_ doesn't have much to do with the interpolated texture coordinates however, but _pIndex_ does. It's badly named, but it's essentially a value between 0 and 2. 0 means that the attributes from the first vertex will be kept, 1 means that the attributes from the second vertex will be kept, and 2 means that they will be interpolated half-way (linear interpolation 50%).

Would you mind explaining what the problematic result is? Is it texture bleeding, or warped coordinates? Do you have a screenshot or a sample mesh that you could share?

**Comment by blair-ahlquist** [comment]
> 2 means that they will be interpolated half-way (linear interpolation 50%).

I don't think this is what should happen. It results in textures being warped:

![abbeyroad](https://user-images.githubusercontent.com/5587927/53686556-7206d480-3cdd-11e9-881c-fcd3f2839457.PNG).

I wrote a test that made a flat grid mesh (z = 0, and x and y ranging from 0 to 1) with texture coordinates identical to the x and y of the vertex coordinates. My assumption is that the texture coordinates after the reduction will still be (nearly) identical to the vertex coordinates so that the texture appearance is preserved. I fixed this in a fork by getting the barycentric coordinates of `p` within the triangle and using that to weight the attributes from all three vertices of the triangle.

**Comment by Whinarn** [comment]
> My assumption is that the texture coordinates after the reduction will still be (nearly) identical to the vertex coordinates so that the texture appearance is preserved.

This would be my assumption as well. I have noticed that the texture coordinates would sometimes get scewed. But what you are saying totally makes sense. I'm not sure why I never thought of that. Let me do some testing on this as well. Thank you!

**Comment by Whinarn** [comment]
I just tested it with a flat grid mesh, just as you described, and I can not see any warped texture coordinates like what you are showing (without any modifications to the code).

**After decimation**
![image](https://user-images.githubusercontent.com/2932594/53690992-94215500-3db8-11e9-8c20-7239578d9d00.png)
**Before decimation**
![image](https://user-images.githubusercontent.com/2932594/53691041-933cf300-3db9-11e9-8c70-a1b201ea34ac.png)

I am doing the testing through Unity however, but that shouldn't affect the decimation unless there is a problem elsewhere in the code?

I still want to test your suggested fix however. So I would like to get hold of a mesh that ends up with warped texture coordinates. Is your grid not uniform or something? I am trying to figure out why we would get such different results. Are you using the latest version of MeshDecimator?

**Comment by Whinarn** [comment]
In the commit https://github.com/Whinarn/MeshDecimator/commit/a2f626c87f59ff1279680d4b905f3aedbaae65a7 of branch [fix/barycentric-attribute-interpolation](https://github.com/Whinarn/MeshDecimator/tree/fix/barycentric-attribute-interpolation) I implemented vertex attribute interpolation using the barycentric coordinates within the triangles during the decimation. And in my personal tests I have noticed a slight improvement. Is this what you had in mind?
